### PR TITLE
Reconstruct go ps: add tensor definition

### PR DESCRIPTION
**Pull request by lwpyr**

### CODE REVIEW

File: elasticdl/pkg/common_new/tensor.go [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+	if int(DimProduct(dims))*int(DtypeSize[pb.GetDtype()]) != len(pb.GetTensorContent()) {
+		return nil
+	}
+
+	var t = Tensor{
+		Buffer: pb.GetTensorContent(),
+		Dims:   dims,
+		Dtype:  pb.GetDtype(),
+	}
+	return &t
```
Comment (QiJune) [inline elasticdl/pkg/common_new/tensor.go]: How about
 
```go
return &Tensor {
  Buffer: pb.GetTensorContent(),
  Dims:   dims,
  Dtype:  pb.GetDtype(),
}

```

File: elasticdl/pkg/common_new/tensor.go [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+	}
+	pbDim := tensor_shape_go_proto.TensorShapeProto{
+		Dim: shapeDim,
+	}
+	var pb = tensor_go_proto.TensorProto{
+		TensorContent: t.Buffer,
+		TensorShape:   &pbDim,
+		Dtype:         t.Dtype,
+	}
+	return &pb
```
Comment (QiJune) [inline elasticdl/pkg/common_new/tensor.go]: The same

```go
return &tensor_go_proto.TensorProto {...}
```

File: elasticdl/pkg/common_new/tensor.go [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+// SubTensor get the part reference of the tensor
+func SubTensor(t *Tensor, begin int64, length int64) *Tensor {
+	dsize := int64(DtypeSize[t.Dtype])
+	begin *= dsize
+	var subt = Tensor{
+		Buffer: t.Buffer[begin : begin+length*dsize],
+		Dims:   []int64{length},
+		Dtype:  t.Dtype,
+	}
+	return &subt
```
Comment (QiJune) [inline elasticdl/pkg/common_new/tensor.go]: The same:
```go
return &Tensor {...}
```

File: elasticdl/pkg/common_new/tensor.go [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+// SetSubTensor set a vector to an index of tensor
+func SetSubTensor(t *Tensor, begin int64, length int64, val *Tensor) {
+	dsize := int64(DtypeSize[t.Dtype])
+	begin *= dsize
+	length *= dsize
+	copy(t.Buffer[begin:begin+length], val.Buffer)
+}
+
+// SetTensorRow set a vector to an index of tensor
+func SetTensorRow(t *Tensor, idx int64, vec *Tensor) {
```
Comment (QiJune) [inline elasticdl/pkg/common_new/tensor.go]: These four functions could be taken as the memeber function of struct `Tensor`. How about following definition:

```go
func (t* Tensor) GetSubTensor(begin int64, length int64) *Tensor
func (t *Tensor) SetSubTensor(begin int64, length int64, val *Tensor)

func (t* Tensor) GetRow(idx int64) *Tensor
func (t* Tensor) SetRow(idx int64, vec *Tensor)
```

Comment (lwpyr) [reply elasticdl/pkg/common_new/tensor.go]: done

File: elasticdl/pkg/common_new/tensor.go (hunk 3 above)
Comment (lwpyr) [reply elasticdl/pkg/common_new/tensor.go]: done

File: elasticdl/pkg/common_new/tensor.go (hunk 2 above)
Comment (lwpyr) [reply elasticdl/pkg/common_new/tensor.go]: done

File: elasticdl/pkg/common_new/tensor.go (hunk 1 above)
Comment (lwpyr) [reply elasticdl/pkg/common_new/tensor.go]: done

File: elasticdl/pkg/common_new/tensor_test.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+	assert.Equal(t, Slice(t1).([]float32), slice, "NewTensor FAIL")
+	assert.Equal(t, t1.Dims, dim, "NewTensor FAIL")
+
+	v1 := NewVector(slice)
+	assert.Equal(t, Slice(v1).([]float32), slice, "NewVector FAIL")
+	assert.Equal(t, []int64{6}, v1.Dims, "NewVector FAIL")
+
+	r1 := t1.GetRow(1)
+	assert.Equal(t, Slice(r1).([]float32), slice[3:6], "RowOfTensor FAIL")
+	assert.Equal(t, []int64{3}, r1.Dims, "RowOfTensor FAIL")
```
Comment (QiJune) [inline elasticdl/pkg/common_new/tensor_test.go]: RowOfTensor FAIL --> GetRow FAIL

File: elasticdl/pkg/common_new/tensor_test.go [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+	val := NewVector([]float32{30, 40, 50})
+	t1.SetRow(1, val)
+	assert.Equal(t, Slice(t1).([]float32), []float32{1.0, 2.0, 3.0, 30, 40, 50}, "SetTensorRow FAIL")
```
Comment (QiJune) [inline elasticdl/pkg/common_new/tensor_test.go]: SetTensorRow FAIL --> SetRow FAIL

File: elasticdl/pkg/common_new/tensor_test.go (hunk 5 above)
Comment (lwpyr) [reply elasticdl/pkg/common_new/tensor_test.go]: done

File: elasticdl/pkg/common_new/tensor_test.go (hunk 6 above)
Comment (lwpyr) [reply elasticdl/pkg/common_new/tensor_test.go]: done

**Review by QiJune — APPROVED** [review]
Excellent!

**Pull request merged**